code/app/src/main/java/com/example/sigma_blue/activities/GalleryActivity.java

```diff
@@ -0,0 +1,221 @@
+    }
+
+    private void StartGallery(){
+        Intent galleryPhotoPick  = new Intent(Intent.ACTION_PICK, android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
+        try {
+            startActivityForResult(galleryPhotoPick , REQUEST_GALLERY_PICKING);
+        } catch (ActivityNotFoundException e) {
+            // display error state to the user
+        }
+    }
+
+    private void checkAndroidStoragePermissions() {
+        int permissionStatus = ContextCompat.checkSelfPermission(this.getApplicationContext(), Manifest.permission.READ_EXTERNAL_STORAGE);
+
+        storagePermissionGranted = (permissionStatus == 0);
+
+        if (!storagePermissionGranted) {
+            // need to ask for permission
+            String[] p = {android.Manifest.permission.READ_EXTERNAL_STORAGE};
+            this.requestPermissions(p, 1);
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int r, String[] p, int[] g) {
+        super.onRequestPermissionsResult(r, p, g);
+        if (g.length > 0 && g[0] == 0) {
+            // If the storage permission is successfully granted
+            storagePermissionGranted = true;
+            StartGallery();
+        } else {
+            // If the user doesn't want out app to use the camera go back to the edit page
+            globalContext.newState(ApplicationState.EDIT_ITEM_FRAGMENT);
+
+            finish();
+        }
+    }
+
+    //
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+
+        super.onActivityResult(requestCode, resultCode, data);
+        Bundle extras = data.getExtras();
+
+        if (requestCode == REQUEST_GALLERY_PICKING && resultCode == RESULT_OK) {
+
+            // extract image from path of image, then upload to the
+            Uri selectedImage = data.getData();
+            String[] filePathColumn = {MediaStore.Images.Media.DATA};
+            if (selectedImage != null) {
+                Cursor cursor = getContentResolver().query(selectedImage, filePathColumn, null, null, null);
+                if (cursor != null) {
+                    cursor.moveToFirst();
+                    int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
+                    String picturePath = cursor.getString(columnIndex);
+
+                    //decode and compress the image to target for uploading and displaying
+                    Bitmap imageBitmap = compressBitmap(picturePath, 1000, 1000);
+                    cursor.close();
+
+                    String path = globalContext.getImageManager().uploadImage(globalContext.getAccount(), imageBitmap);
+                    globalContext.getModifiedItem().addImagePath(path);
+                    globalContext.newState(globalContext.getLastState());
+                    finish();
+                }
+            }
+
+
+        }
+    }
+
+    // bitmap compressing method
+    public Bitmap compressBitmap(String path, int reqWidth, int reqHeight) {
```

Comment:
I think this should be somewhere else, it's not really a UI thing. Maybe we need an image class? For now it could go to the imageDB class.